File: heap_sort.cpp
//{ Driver Code Starts
// C++ program for implementation of Heap Sort
#include <bits/stdc++.h>
using namespace std;

// } Driver Code Ends
// The functions should be written in a way that array become sorted
// in increasing order when heapSort() is called.

class Solution {
 public:
  // Heapify function to maintain heap property.
  void heapify(int arr[], int n, int i) {
    // Your Code Here
    while (2 * i + 1 < n) {
      int j = 2 * i + 1;
      if (j + 1 < n && arr[j] < arr[j + 1]) ++j;
      if (arr[i] < arr[j]) {
        swap(arr[i], arr[j]);
        i = j;
      } else {
        break;
      }
    }
  }

 public:
  // Function to build a Heap from array.
  void buildHeap(int arr[], int n) {
    // Your Code Here
    for (int i = n / 2; i >= 0; --i) {
      heapify(arr, n, i);
    }
  }

 public:
  // Function to sort an array using Heap Sort.
  void heapSort(int arr[], int n) {
    // code here
    buildHeap(arr, n);
    while (n > 1) {
      swap(arr[0], arr[n - 1]);
      --n;
      heapify(arr, n, 0);
    }
  }
};

//{ Driver Code Starts.

/* Function to print an array */
void printArray(int arr[], int size) {
  int i;
  for (i = 0; i < size; i++) printf("%d ", arr[i]);
  printf("\n");
}

// Driver program to test above functions
int main() {
  int arr[1000000], n, T, i;
  scanf("%d", &T);
  while (T--) {
    scanf("%d", &n);
    for (i = 0; i < n; i++) scanf("%d", &arr[i]);
    Solution ob;
    ob.heapSort(arr, n);
    printArray(arr, n);
  }
  return 0;
}

// } Driver Code Ends

File: sort/heap_sort.py
# User function Template for python3


class Solution:
    # Heapify function to maintain heap property.
    def heapify(self, arr, n, i):
        # code here
        while i * 2 + 1 < n:
            child = i * 2 + 1
            if child + 1 < n and arr[child + 1] > arr[child]:
                child += 1
            if arr[i] > arr[child]:
                break
            arr[i], arr[child] = arr[child], arr[i]
            i = child

    # Function to build a Heap from array.
    def buildHeap(self, arr, n):
        # code here
        for i in range(n // 2 - 1, -1, -1):
            self.heapify(arr, n, i)

    # Function to sort an array using Heap Sort.
    def HeapSort(self, arr, n):
        # code here
        self.buildHeap(arr, n)
        for i in range(n - 1, -1, -1):
            arr[0], arr[i] = arr[i], arr[0]
            self.heapify(arr, i, 0)


# {
# Driver Code Starts
# Initial Template for Python 3
import atexit
import io
import sys

# Contributed by : Mohit Kumara

_INPUT_LINES = sys.stdin.read().splitlines()
input = iter(_INPUT_LINES).__next__
_OUTPUT_BUFFER = io.StringIO()
sys.stdout = _OUTPUT_BUFFER


@atexit.register
def write():
    sys.__stdout__.write(_OUTPUT_BUFFER.getvalue())


if __name__ == "__main__":
    test_cases = int(input())
    for cases in range(test_cases):
        n = int(input())
        arr = list(map(int, input().strip().split()))
        Solution().HeapSort(arr, n)
        print(*arr)

# } Driver Code Ends
